### PR TITLE
Extend comments API

### PR DESCRIPTION
**Pull request by msobon-hexrays**
Added new APIs:

- `append_at(ea, comment, comment_kind=REGULAR)` - Appends text to a regular or repeatable comment on a new line, creating the comment if missing. Always works on the item comment (function comments are managed by the functions API).
- `set_extra_lines_at(ea, lines, kind)` - Replaces all anterior or posterior comment lines with the given iterable.
- `delete_extra_lines_at(ea, kind)` - Deletes all anterior or posterior comment lines, including lines hidden behind index gaps.
- `get_combined_at(ea, *, include_regular, include_repeatable, include_anterior, include_posterior)` - Returns the selected comments at an address joined into one newline-separated block, in listing order. Matching the listing behavior, the repeatable comment is included only when no regular comment is.

### CODE REVIEW

File: ida_domain/comments.py [hunk 1]
```diff
@@ -160,6 +173,52 @@ def delete_at(self, ea: int, comment_kind: CommentKind = CommentKind.REGULAR) ->
...
+            True if the comment was successfully appended, False otherwise.
+
+        Warning:
+            IDA caps regular and repeatable comments at 1024 bytes of UTF-8
+            and silently truncates longer text, so appending to a comment
+            near that limit drops the excess even though True is returned.
+        """
+        if not self.database.is_valid_ea(ea):
+            raise InvalidEAError(ea)
+        if ida_bytes.is_tail(ida_bytes.get_flags(ea)):
```
Comment (bbudny-hexrays) [inline ida_domain/comments.py]: can use `self.database.bytes.is_tail_at` instead. but additionally, you check `is_tail` here while, in `_set_extra_lines_at` and it's friends you don't (preexisting `set_at` `get_at` etc functions don't really care either, will happily "operate" on a tail byte in various ways) - inconsistent behavior, decide on one failure/success mode; i think we should follow the behavior of preexisting implementations + document the quirks (incl. in preexisting functions such as `set_at` which returns `True` but  doesn't display it [yet still sets it in the idb :/] ; no mention of that in the docstring).

File: ida_domain/comments.py [hunk 2]
```diff
@@ -160,6 +173,52 @@ def delete_at(self, ea: int, comment_kind: CommentKind = CommentKind.REGULAR) ->
...
+            raise InvalidParameterError('ea', ea, 'the address must not be an item tail byte')
+        if comment_kind == CommentKind.ALL:
+            raise InvalidParameterError(
+                'comment_kind',
+                comment_kind,
+                'appending to both comment kinds at once is not supported',
+            )
+        if not isinstance(comment, str):
```
Comment (bbudny-hexrays) [inline ida_domain/comments.py]: perhaps we should also check `isinstance` of `comment_kind` against `CommentKind`, otherwise if somebody passes a non-enum value (e.g. `True` or `False`, perhaps old habit from idapython funcs) the function will happily accept it and always put a non-repeatable comment due to `comment_kind == CommentKind.REPEATABLE` == `False` . On the other hand, similarly to tail-bytes, the preexisting implementations don't check this kind of stuff at all so for consistency reasons the new ones probably shouldn't either.

File: ida_domain/comments.py [hunk 3]
```diff
@@ -160,6 +173,52 @@ def delete_at(self, ea: int, comment_kind: CommentKind = CommentKind.REGULAR) ->
...
+            raise InvalidParameterError('comment', comment, 'the comment must be a string')
+        if not comment:
+            return False
```
Comment (bbudny-hexrays) [inline ida_domain/comments.py]: arguably if a user calls `append_at` with an empty string, they successfully  "appended nothing" so i'm not sure if a false return value is justified here

File: ida_domain/comments.py [hunk 4]
```diff
@@ -160,6 +173,52 @@ def delete_at(self, ea: int, comment_kind: CommentKind = CommentKind.REGULAR) ->
...
+
+        return ida_bytes.append_cmt(ea, comment, comment_kind == CommentKind.REPEATABLE)
```
Comment (bbudny-hexrays) [inline ida_domain/comments.py]: at a function head address this will interact with the function comment (`append_cmt` goes to `get`/`set_func_or_item_cmt`), while `set_at`, `get_at` et al. only interact with the underlying item (i.e. they bypass the function comment and go straight to the instruction)  - inconsistent design.  `append_at` should only work on item comments like `set_cmt` i reckon. maybe also adding test cases for behavior at function start boundaries wouldn't hurt

File: ida_domain/comments.py [hunk 5]
```diff
@@ -160,6 +173,52 @@ def delete_at(self, ea: int, comment_kind: CommentKind = CommentKind.REGULAR) ->
...
 
+    def append_at(
+        self, ea: ea_t, comment: str, comment_kind: CommentKind = CommentKind.REGULAR
+    ) -> bool:
+        """
+        Appends text to a comment at the specified address.
+
+        If the selected comment does not exist, it is created with the given
+        text. Otherwise, a newline followed by the supplied text is appended.
+        Appending an empty string is a no-op and returns False.
```
Comment (bbudny-hexrays) [inline ida_domain/comments.py]: see below

File: tests/test_comments.py
Comment (bbudny-hexrays) [inline tests/test_comments.py:1]: maybe wouldn't hurt to add some `get_dissassembly` tests to make sure our changes to comments actually get rendered, instead of just writing to the netnode or sth

**Review by bbudny-hexrays — CHANGES_REQUESTED** [review]

File: ida_domain/comments.py [hunk 6]
```diff
@@ -216,6 +275,155 @@ def set_extra_at(self, ea: int, index: int, comment: str, kind: ExtraCommentKind
...
+            raise InvalidEAError(ea)
+
+        base_idx = ida_lines.E_PREV if kind == ExtraCommentKind.ANTERIOR else ida_lines.E_NEXT
+        success = True
+        for index in range(base_idx, base_idx + _EXTRA_COMMENT_SLOT_COUNTS[kind]):
+            if ida_lines.get_extra_cmt(ea, index) is not None:
+                success = ida_lines.del_extra_cmt(ea, index) and success
+        return success
+
+    def set_anterior_lines_at(self, ea: ea_t, lines: Iterable[str]) -> bool:
```
Comment (cmghex) [inline ida_domain/comments.py]: Currently, the existing extra-comment method takes the anterior/posterior choice as a parameter called `kind`:
- db.comments.set_extra_at(ea, 0, 'hello', ExtraCommentKind.ANTERIOR)
- db.comments.get_extra_at(ea, 0, ExtraCommentKind.ANTERIOR)
- db.comments.get_all_extra_at(ea, ExtraCommentKind.ANTERIOR)
- db.comments.delete_extra_at(ea, 0, ExtraCommentKind.ANTERIOR)

Here, we have the choice into the method name, so each operation needs two methods:
- db.comments.set_anterior_lines_at(ea, lines)     # anterior
- db.comments.set_posterior_lines_at(ea, lines)    # posterior
- db.comments.clear_anterior_at(ea)                # anterior
- db.comments.clear_posterior_at(ea)               # posterior

Could we rather use the same convention for uniformity? Pass a `kind` parameter (maybe make that enumeration more generic and reuse it also here)

File: ida_domain/comments.py [hunk 7]
```diff
@@ -216,6 +275,155 @@ def set_extra_at(self, ea: int, index: int, comment: str, kind: ExtraCommentKind
...
+            a line fails, the original lines are already lost and only a
+            partial replacement remains stored.
+
+        Warning:
+            IDA caps each line at 1024 bytes of UTF-8 and silently truncates
+            longer lines even though True is returned.
+        """
+        return self._set_extra_lines_at(ea, lines, ExtraCommentKind.POSTERIOR)
+
+    def clear_anterior_at(self, ea: ea_t) -> bool:
```
Comment (cmghex) [inline ida_domain/comments.py]: Also, as naming convention/uniformity, for the other removals we are using "delete_<>_at"

File: ida_domain/comments.py [hunk 8]
```diff
@@ -263,6 +471,59 @@ def get_all_extra_at(self, ea: int, kind: ExtraCommentKind) -> Iterator[str]:
             yield comment
             index += 1
 
+    def get_combined_at(
```
Comment (cmghex) [inline ida_domain/comments.py:415]: get_all_at(ea, …) would pair naturally with get_at/get_all, don't you think?

File: ida_domain/comments.py (hunk 1 above)
Comment (msobon-hexrays) [reply ida_domain/comments.py]: Replaced with `self.database.bytes.is_tail_at`.

Regarding the inconsistency in checking the `is_tail`, this is on purpose, as the underlying IDA behavior is different. For setting regular/repeatable comments IDA discards anything that is not set in the `head` item and does not store them in the IDB, while for extra lines it does store them in IDB even for non-head items. 

I think it can be useful in some cases for plugins (to set a comment on an arbitrary address without caring if this is a head or not and be able to retrieve it later), and the behavior is consistent with idapython, but indeed, in the disassembly view they are rendered only for head items, so we need to decide whether we want to allow such comments or not.

I agree the behavior should be consistent with `set_at`, but I'd rather add the check for tail there instead of extending the docs, but as it is a breaking change I would do it in separate PR.

For `get_at` I think current behavior is ok. ida-python accepts non-head bytes and normalize them to head. So this is consistent with ida-python

File: ida_domain/comments.py (hunk 2 above)
Comment (msobon-hexrays) [reply ida_domain/comments.py]: Removed the `isinstance(comment, str):` to be consistent

File: ida_domain/comments.py (hunk 3 above)
Comment (msobon-hexrays) [reply ida_domain/comments.py]: I don't think anyone ever appends empty string to a comment on purpose as it doesn't make any sense, so wanted to make some check to help catch potential mistakes, but removed for now to keep it consistent.

File: ida_domain/comments.py (hunk 4 above)
Comment (msobon-hexrays) [reply ida_domain/comments.py]: Fixed

File: tests/test_comments.py
Comment (msobon-hexrays) [reply tests/test_comments.py:1]: Added a test case

File: ida_domain/comments.py (hunk 7 above)
Comment (msobon-hexrays) [reply ida_domain/comments.py]: Fixed

File: ida_domain/comments.py (hunk 8 above)
Comment (msobon-hexrays) [reply ida_domain/comments.py:415]: The name sounds nice but the thing is that both `get_at` and `get_all` return comments as they are stored, while  `get_combined_at` returns a single string that is combines comment of all comment types, so I would say this is kinda artificial (this is just to match the way comments are displayed in disassembly).

I fixed one more thing there (I noticed when both regular and repeatable commits are set, IDA displays only regular ones)

File: ida_domain/comments.py (hunk 6 above)
Comment (msobon-hexrays) [reply ida_domain/comments.py]: Right, updated

**Review by adybov-hexrays — APPROVED** [review]

File: ida_domain/comments.py (hunk 1 above)
Comment (bbudny-hexrays) [reply ida_domain/comments.py]: i see, though it seems a little weird to me to reinforce what appears to be an inconsistency in underlying behavior (compatibility with idapython doesn't really speak to me to be honest), and i don't think it's an intuitive thing. could it be useful? i guess so, not sure for what exactly, but this is "hacks" territory for me.
i would personally just reject comments that don't get displayed to set a precedent, but if at least something gets written to the IDB then that's ok to me too. agreed about `set_at`, and `get_at`, works for me.
if you disagree with me thesis, feel free to consider this resolved (resolve the comment yourself, if i resolve it right now you won't see it, if you don't resolve it you won't merge).

**Review by bbudny-hexrays — APPROVED** [review]
